### ISSUE TITLE
Promise CE is executed multiple times if there are multiple subscribers

### ISSUE THREAD

**Issue by 0x53A** [fable-compiler/fable-powerpack#59]
If I create and cache a promise, then it is re-executed each time I do a ``.then``.

A reproduction is [here](https://fable.io/repl/#?code=LYewJgrgNgpgBAUSsAlgZwBYDoDKLgAOsAkgHYEQAuAUNSATKXAGICGARrFgMIgBOMLACk0ZSjD706DJm06CAEjCgM+aLACUYrAMY16jFhy5KVE9Vt2UsABUkE00w0lSYnTF+myW97o-NsQAHcJG10Aa1pKAE8GOAAVOABeOABvOAVWTDgALjg0Sj4UUgBzOABfWlhKOFYoKAA1OogYNGTqOE64AG0AHw6uwfTM7JSAIgAGAgBpAFYCAA0ATSCAMxQALQBOAEUARwBmSgA2AHUARgBVAA8IY4AmGPiAagwoAHpp6bAFkAAve5CeIHYBLJJjCoDQadYZZDDJOCTGbzZZrTa7Q4nC43O6PaIvN6fb6-AFAkFgiGVaFdWGjRFTOaLFbrbb7I5nK63B5PV4fL4-f6A4Gg8GQ6kwjJwhFIxmolkY9nYrl4gl84mCskiylQ6G0+HjBko5notlYzm4nmE-kkoXk0VUrq9AC6VRgNTA6EoxT0TSgLTaSR1fR1gxlRrRrMxHJx3PxvKJAtJwopOud1AIRVIlFWTDGAFIUHAPQVvTUAG7NVoQ4te0g+yvqAAyjBKlAwrpqKDAAFlWAR2oMa6Xff7UwA+OA4GB7LCUEAAQT4fFY0XHcEXy+iWGAfbgAApVhAmF24ABaMchrrVOAnlJdy+da-Adj9lIZkCueCpB-Q6+3m9gD+gwZsU2a5gAIiAxRlHmaAQve4rQgIlAQHwx6AYhlQ-u87xwHOcDrNcN6UAANHAR46B+wCMDUbboHAUDFDAOTYe8T4vgiz4EFgAAGPFtowfEHkecBEeeokAJS0IhXZkVxOpSYMvQTswOAYKwfDcbw9QwHoKAgKQ6i9txICrFOey0Ks-BwMAN5MHUjQNkWIA6teAD676fmACKeegX4-tUACENnsAiXbGVg3TAFgIwYC6iF7opXQOp0chcDYwShBEtiSJ+WAFBpNQeblflgEAA&html=DwCwLgtgNgfAUHUBTAhgE3gAm54ElgqbhgAOAtEgI4CuAlgG4C8A5AMID2AdmEj+QBUAnqSQtMAY268erXgA8wAenDQA3JJAoATgGcCTGmABm5ABwssOYLona6pMJl3aJTAEQB9TwAkA8gDKAt5KUHQARrpK2qgSYAB0pNocaDRxdNzxEHRc8QBWuu4wwEq29o5W2DZ2Dk4ubl6+gcGeoRFRMShx5GgcEInJqemZ2bkFRSVltfAlIKgYCMDhKUKSUCi6uh4opBRSPCg5SNpFcDi4aIyYdGgeSFDZuiDkO6Tuaxtb7vePz68TSkuDBmSmWaCE8EQKkgsDgQA&css=BYFwtgNgNAUARgewCYE8AEBvGadoGYIB2IAtAM4CWAXgKYBcaAjAGwAOAHgNwwC+MMAOjABDCoRIBjIiFGEaAJ0zZcSCmVYRhKBngg0uynAHcKSEMAaMADFYCk3XGmA0KAc1CWb9w2gBWAVzIQCjwUSWkaYgYJSJAFB1xhCDdxCjiwMmjY+J9dfRJVeRoJYKJohAh-MEJuPkExVn8QJUdWYSRVQlcGAQAmAFYisAScAmJyanomNgNHEzMLNAGrDhG0EXlXMRJEEBAEMEsh2pggA).

In that snippet, I create a ``Map<string, Promise<string>>``.

I would expect the promise to behave like the C# ``async``, be executed once, and further ``let!`` to access the cached value.

Instead it behaves like the F# ``async`` and is re-executed each time.

**Comment by 0x53A** [comment]
Simpler repro:

```f#
module Elmish.SimpleInput

open Fable.Core.JsInterop
open Fable.Helpers.React
open Fable.Helpers.React.Props
open Elmish
open Elmish.React
open Fable.PowerPack

let p = promise {
    printfn "Inside Promise"
    return 1
}


Fable.PowerPack.Promise.start( p.``then``(fun i ->
    printfn "Then 1"
) )

Fable.PowerPack.Promise.start( p.``then``(fun i ->
    printfn "Then 2"
) )

Fable.PowerPack.Promise.start( p.``then``(fun i ->
    printfn "Then 3"
) )
```

![grafik](https://user-images.githubusercontent.com/4236651/48723613-5e03ff00-ec27-11e8-9384-e53ec0651765.png)

**Comment by MangelMaxime** [comment]
I think this is normal, because if I am not wrong promise is executed/triggered by `then`

**Comment by 0x53A** [comment]
AFAIK it is triggered by ``then``, but subsequent executions do not re-trigger it.

If I run this in JSFiddle:

```js
var p1 = new Promise(
    function(resolve, reject) {       
      console.log("Inside Promise");
      
      resolve(1);
  }
);
    
var p2 = p1.then(function(val) {
  console.log("Inside then 1");
});

 
var p3 = p1.then(function(val) {
  console.log("Inside then 2");
});
```

I get this output:

![grafik](https://user-images.githubusercontent.com/4236651/48725513-b7baf800-ec2c-11e8-8eef-876a6764f762.png)

Which is what I would expect.

**Comment by alfonsogarciacaro** [comment]
Yes, this is kind of intended. It can be debatable if the `promise CE` should behave like JS Promise (hot) or F# Async (cold) but in any this is how F# computation expressions work: they create a computation and in the desugared code the whole expression is wrapped in a `delay` function that is rerun every time. Not sure if it's possible to change this behaviour (don't remember if I tried).

If you need a hot promise you can use [Promise.create](https://github.com/fable-compiler/fable-powerpack/blob/3d0ef40d9c4969e423361443f6ae5daa2244fc11/src/Promise.fs#L15-L17).

**Comment by 0x53A** [comment]
I would prefer if it behaved as much like the native JS promise as possible, I was wondering for a long time why my client fired so many requests even though I added such a nice deduplication and cache ¯\\_(ツ)_/¯.

Afaik Fable also supports the ``async`` CE? Then the user even has a choice - use ``async`` if you want cold, use ``promise`` if you want pseudo-hot (since it doesn't start immediately, but then gets cached).

Question is, do you agree with me? And, since Fable is now in version 2 and (relatively) stable, would such a breaking change be accepted? Or maybe for Fable 3? ;D

(I haven't looked, but am pretty sure that _technically_ the CE can be changed. I can take a look tomorrow.)

**Comment by alfonsogarciacaro** [comment]
Fable.PowerPack was removed from core intentionally so we can make these breaking changes. Personally, I'd probably be ok with this change, but I'm not sure if other people are already relying on this behaviour or find it more intuitive. Let's see if we can get some feedback.

**Comment by alfonsogarciacaro** [comment]
Fixed by #61
